[Cleanup] remove masternode broadcast from CActiveMasternode class

random-zebra opened the pull request:
Based on top of
- [x] #1698 

This removes unused/duplicated code in `CActiveMasternode` (ref: https://github.com/PIVX-Project/PIVX/pull/1698#pullrequestreview-439245589).
It also removes the RPC `masternodedebug`, which is a buggy simplified version of `getmasternodestatus`.

furszy commented on src/activemasternode.cpp:
```diff
@@ -86,51 +86,6 @@ void CActiveMasternode::ManageStatus()
-
-        if (GetMasterNodeVin(vin, pubKeyCollateralAddress, keyCollateralAddress)) {
-            if (GetInputAge(vin) < MASTERNODE_MIN_CONFIRMATIONS) {
-                status = ACTIVE_MASTERNODE_INPUT_TOO_NEW;
```
with this removal, this state `ACTIVE_MASTERNODE_INPUT_TOO_NEW ` will never be set.

If it's not used anywhere, we can remove it from the states list.

random-zebra replied:
Actually, I think it would be better to restore the age check (it's useful for the operator in that case).

random-zebra commented:
Rebased on master and added two commits:
- e6a1a51fb52ead330f426b553aa3c9d01a84fe33 refactors `CActiveMasternode::vin` using the `Optional<>` class, as the input is set only during `EnableHotColdMasternode()` (either upon receiving the broadcast or during update). So having the vin un-initialized is a valid state, and we should check it against `nullopt` where it's used.
- 1365ee91df71f422e4a886341fa3c0f50c3a0405 addresses @furszy's comment, removing the `ACTIVE_MASTERNODE_INPUT_TOO_NEW` state.

furszy commented:
Has a small conflict.

random-zebra commented:
Rebased.

furszy commented on src/activemasternode.h:
```diff
@@ -35,14 +35,15 @@ class CActiveMasternode
 
     CActiveMasternode()
     {
+        vin = nullopt;
```
would be better to move it to the direct member initializer.

furszy commented on src/activemasternode.cpp:
```diff
@@ -113,6 +113,11 @@ std::string CActiveMasternode::GetStatus()
 
 bool CActiveMasternode::SendMasternodePing(std::string& errorMessage)
 {
+    if (!vin.is_initialized()) {
```
let's better compare it against `nullopt` to know whether is initialized or not. 

The main goal of the Optional class is to move away from boost internal methods, to make smoother the transition to c++17 std::optional.

random-zebra replied:
fair enough.

furszy commented on src/masternode-budget.cpp:
```diff
@@ -2208,7 +2216,7 @@ void CFinalizedBudget::SubmitVote()
     }
 
-    CFinalizedBudgetVote vote(activeMasternode.vin, GetHash());
+    CFinalizedBudgetVote vote(activeMasternode.vin.get(), GetHash());
```
Better to change `activeMasternode.vin.get()` for `*activeMasternode.vin`

random-zebra commented:
Those are not necessarily related to the current cleanup. The getBalance output, for example, is used to check whether this is a hot or cold masternode, and to remove isLocked, we should first ensure that there's no other part in the masternode flow requiring an unlocked wallet.
I think that the `CActiveMasternode::ManageStatus` method would benefit from a complete overhaul, but that's out of scope for this particular PR, which is just for the removal of the masternode broadcast code duplication.
The whole flow and relative class will be changed with deterministic masternodes anyway.

furszy commented:
Ok np on doing it in another PR, impossible to disagree on the complete overhaul. Went into this direction because we removed the "automatic" hot MN startup on this PR.
